File: Kattis/solutions/cardTrick2/cardTrick2.cpp
#include <iostream>
#include <list>
#include <vector>
using namespace std;

int main() {
  int n;
  cin >> n;

  vector<list<int>> res;
  for (int i = 1; i <= 13; i++) { // Generate all answers
    list<int> list;
    for (int j = i; j > 0; j--) {
      list.push_front(j);
      for (int k = 0; k < j; k++) {
        int temp = list.back();
        list.pop_back();
        list.push_front(temp);
      }
    }
    res.push_back(list);
  }

  for (int i = 0; i < n; i++) {
    int query;
    cin >> query;

    for (auto& num : res[query - 1]) {
      cout << num << " ";
    }
    cout << "\n";
  }

  return 0;
}

File: Kattis/solutions/cardTrick2/cardTrick2.java
import java.util.*;

public class cardtrick2 {
  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    int n = scanner.nextInt();

    ArrayList<LinkedList<Integer>> res = new ArrayList<>();
    for (int i = 1; i <= 13; i++) {
      LinkedList<Integer> list = new LinkedList<>();
      for (int j = i; j > 0; j--) {
        list.addFirst(j);
        for (int k = 0; k < j; k++) {
          int temp = list.removeLast();
          list.addFirst(temp);
        }
      }
      res.add(list);
    }

    for (int i = 0; i < n; i++) {
      int query = scanner.nextInt();
      for (int num : res.get(query - 1)) {
        System.out.print(num + " ");
      }
      System.out.println();
    }

    scanner.close();
  }
}
